### PR TITLE
fix: warm Muon all-to-all connections during optimizer initialization

### PR DESCRIPTION
**Pull request by samsja**
## Summary

Automatically establish the NCCL peer connections used by Dion's Muon optimizer immediately after constructing Muon and before returning it to the trainer.

The warm-up runs only for Muon. It covers the exact main optimizer process group and, when expert parameters use expert parallelism, the separate expert-shard process group. Single-rank and non-NCCL groups remain no-ops; AdamW, SGD, and other optimizers are unchanged.

This is a required correctness workaround for a reproducible multi-node GLM-4.5-Air first-step deadlock. The source now documents that optimizer refactors must preserve the warm-up, its process groups, and its placement between Muon construction and the first optimizer step.

## Failure signature

Pretrained GLM-4.5-Air training with Muon reproducibly stalled while submitting the first optimizer all-to-all on two nodes. All 16 ranks reached the collective, but only rank 0 enqueued it successfully; ranks 1–15 remained inside submission until the NCCL watchdog fired.

The clearest controlled failure was job 1182:

- Nodes `qpn01gpu009` and `qpn01gpu011`, 16 GPUs total.
- Pretrained eight-layer GLM-Air reproduction with CP4, EP8/automatic expert mesh, BF16 compute, and optimizer-state offload.
- First Muon all-to-all was NCCL sequence 67 with 50,331,648 BF16 elements per rank.
- The watchdog fired after 182 seconds.
- Setting `NCCL_CUMEM_ENABLE=0` did not resolve it.

This was not a generic inability of the nodes to communicate: Adam training and standalone all-reduce/all-to-all controls completed on the same cluster. Fake-weight and reduced-model tests further isolated the trigger to Muon's pretrained-model initialization state and first bulk all-to-all.

## Implementation

`_warmup_muon_mesh` takes the `DeviceMesh`, resolves its single process group, and performs a bulk BF16 list all-to-all matching Dion's communication path, waits for completion, and synchronizes CUDA before releasing its temporary buffers. A scalar collective is intentionally not used because it may not initialize the same bulk-transfer peer channels. Meshes with more than one dimension (the replicate-only world mesh) have no single group and are skipped; Dion's Muon only accepts 1-D meshes there anyway.

The ordering is part of the fix:

1. Construct the Dion Muon optimizer.
2. Warm Muon's main distributed NCCL group.
3. If expert parameters and EP are active, warm the independent expert-shard NCCL group.
4. Return the optimizer only after both required warm-ups complete.

The temporary allocation is approximately 12 MiB per peer for send plus receive buffers. No config flag is introduced because making this optional would allow the known deadlock to recur.

Patch scope remains one optimizer source file. There are no trainer-loop, dependency, precision, optimizer-offload, allocator, Dion, or experiment-config changes.

## Validation

- Ruff lint/format and whitespace checks passed.
- Existing configuration tests passed: 70 passed, 67 deselected.
- Job 1183, reduced pretrained model with early connection warm-up: one full step passed on all 16 ranks; 42.6 seconds, loss 15.4930, gradient norm 49.0578, clean exit.
- Job 1184, full 46-layer pretrained GLM-4.5-Air with the earlier pre-model-load placement: three fake-data steps passed with finite losses and nonzero gradient norms.
- Job 1185, full 46-layer model with the final optimizer-construction placement in this PR: three steps passed on all 16 ranks. Losses were 13.2325, 13.1564, and 13.3399; gradient norms were 27.9833, 27.8885, and 31.8815; clean exit.
- The research-prod Air ablations pin the backport of this implementation. IPO-KL, IPO, DPPO, and IcePop have each subsequently advanced beyond 400 Muon steps without recurrence of the initialization deadlock.
- Review follow-up (`dfa07cd7e`): the warm-up now resolves the process group from the mesh and no-ops on non-1-D meshes. Verified on 2 GPUs with NCCL: the `dp_shard` path still completes the warm-up collectives, and the 2-D replicate-only world mesh returns cleanly where the previous no-arg `get_group()` call raised `RuntimeError`.

Both full-model validations used `qpn01gpu009/011`, 16 GPUs, CP4 + EP8, optimizer-state offload, BF16 compute, and expandable CUDA allocator segments. They used fake data with no evaluation or checkpoint overhead.

Job 1185 artifacts are retained at:

```text
/home/outputs/int4-debug/sft-air-full-muon-init-warmup-2n-009-011
```

## Confidence and limitations

Confidence is high that this initialization prevents the observed first-all-to-all deadlock: the failure was reproducible without it, the matching warm-up removed it in progressively more production-like tests, and the final implementation has continued to work in four long-running RL jobs.

The exact NCCL-native deadlock mechanism remains unproven. The warm-up changes connection state, allocation history, and startup timing, so this should be described as a targeted correctness workaround rather than proof of a particular NCCL defect. Removing or moving it requires reproducing the original multi-node pretrained GLM-Air Muon case, not merely passing a standalone collective test.

---

> [!NOTE]
> **Medium Risk**
> Changes distributed initialization for Muon only, but incorrect group selection or removal could reintroduce multi-node deadlocks or add startup memory/time on every Muon run.
> 
> **Overview**
> Adds a **correctness workaround** for multi-node Muon training: NCCL peer channels for bulk all-to-all are established **right after** Dion `Muon` is built and **before** the first optimizer step, so GLM-Air-style jobs avoid deadlocking on the initial Muon collective.
> 
> New helper `_warmup_muon_process_group` no-ops for single-rank or non-NCCL groups; otherwise it runs an async **list** `all_to_all` on ~3 MiB BF16 buffers per rank (matching Dion’s bulk path), waits, and synchronizes CUDA. Warm-up runs on the main **`dp_shard_cp`** (or world) process group and, when expert MLP params use EP, separately on **`dp_shard_mod_ep`**. AdamW, SGD, and other optimizers are unchanged.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit d0b707438f77780e4b5e0ed69237c08435ba3a41. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: src/prime_rl/trainer/optim/__init__.py [hunk 1]
```diff
@@ -196,4 +214,7 @@ def muon_enabled(n, p):
         world_mesh=parallel_dims.world_mesh,
         fsdp_mesh_dim=1 if parallel_dims.dp_replicate_enabled else 0,
     )
+    _warmup_muon_process_group(distributed_mesh.get_group())
```
[the remark this answers is not in the document]
Comment (samsja) [reply src/prime_rl/trainer/optim/__init__.py]: Addressed in `dfa07cd7e`: the warm-up helper now takes the `DeviceMesh` and resolves the process group itself, no-oping on meshes with `ndim > 1` (the replicate-only world mesh has no single group to warm).

Verified on 2 GPUs with NCCL: the `dp_shard` path still completes the warm-up collectives, and passing the 2-D replicate-only world mesh returns cleanly where the previous no-arg `get_group()` raised `RuntimeError`.

Separate pre-existing note: replicate-only multi-GPU Muon still fails earlier in Dion's constructor (`Only 1D DeviceMesh is supported`) — that crash exists on `main` and is unrelated to the warm-up.

**Review by S1ro1 — APPROVED** [review]
lgtm
